File: example/basics1_display_robot_states.cpp
/**
 * @example basics1_display_robot_states.cpp
 * This tutorial does the very first thing: check connection with the robot server and print
 * received robot states and actions.
 * @copyright Copyright (C) 2016-2025 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/rdk/robot.hpp>
#include <flexiv/rdk/utility.hpp>

#include <iostream>
#include <thread>

using namespace flexiv;

/** @brief Print program usage help */
void PrintHelp()
{
    // clang-format off
    std::cout << "Required arguments: [robot_sn]" << std::endl;
    std::cout << "    robot_sn: Serial number of the robot to connect. Remove any space, e.g. Rizon4s-123456" << std::endl;
    std::cout << "Optional arguments: None" << std::endl;
    std::cout << std::endl;
    // clang-format on
}

/** @brief Print robot states and actions data @ 1Hz */
void PrintRobotStates(rdk::Robot& robot)
{
    while (true) {
        // Print all robot states in JSON format using the built-in ostream operator overloading
        std::cout << "Current robot states:" << std::endl;
        std::cout << robot.states() << std::endl;

        // Print all robot actions in JSON format using the built-in ostream operator overloading
        std::cout << "Current robot actions:" << std::endl;
        std::cout << robot.actions() << std::endl;

        // Print digital inputs
        std::cout << "Current digital inputs:" << std::endl;
        std::cout << rdk::utility::Arr2Str(robot.digital_inputs()) << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

int main(int argc, char* argv[])
{
    // Program Setup
    // =============================================================================================
    // Parse parameters
    if (argc < 2 || rdk::utility::ProgramArgsExistAny(argc, argv, {"-h", "--help"})) {
        PrintHelp();
        return 1;
    }
    // Serial number of the robot to connect to. Remove any space, for example: Rizon4s-123456
    std::string robot_sn = argv[1];

    // Print description
    std::cout << ">>> Tutorial description <<<\nThis tutorial does the very first thing: check "
                 "connection with the robot server and print received robot states and actions.\n"
              << std::endl;

    try {
        // RDK Initialization
        // =========================================================================================
        // Instantiate robot interface
        rdk::Robot robot(robot_sn);

        // Clear fault on the connected robot if any
        if (robot.fault()) {
            std::cerr << "[warn] Fault occurred on the connected robot, trying to clear ..."
                      << std::endl;
            // Try to clear the fault
            if (!robot.ClearFault()) {
                std::cerr << "[error] Fault cannot be cleared, exiting ..." << std::endl;
                return 1;
            }
            std::cout << "Fault on the connected robot is cleared" << std::endl;
        }

        // Enable the robot, make sure the E-stop is released before enabling
        std::cout << "Enabling robot ..." << std::endl;
        robot.Enable();

        // Wait for the robot to become operational
        while (!robot.operational()) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        std::cout << "Robot is now operational" << std::endl;

        // Print States
        // =========================================================================================
        // Use std::thread to do scheduling so that this example can run on all OS, since not all OS
        // support rdk::Scheduler
        std::thread low_priority_thread(std::bind(PrintRobotStates, std::ref(robot)));

        // Properly exit thread
        low_priority_thread.join();

    } catch (const std::exception& e) {
        std::cerr << "[error] " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: example_py/basics1_display_robot_states.py
#!/usr/bin/env python

"""basics1_display_robot_states.py

This tutorial does the very first thing: check connection with the robot server and print
received robot states.
"""

__copyright__ = "Copyright (C) 2016-2026 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import time
import argparse
import threading
import spdlog  # pip install spdlog
import flexivrdk  # pip install flexivrdk


def print_robot_states(robot, logger, stop_event):
    """
    Print robot states data @ 1Hz.

    """

    while not stop_event.is_set():
        # Print available joint groups
        joint_groups_str = " ".join(
            [f"[{flexivrdk.kJointGroupNames[group]}]" for group in robot.groups()]
        )
        logger.info(f"Available joint groups: {joint_groups_str}")

        # Print all robot states in JSON format using the built-in __str__ overloading
        for group, states in robot.states().items():
            logger.info(f"[{flexivrdk.kJointGroupNames[group]}] robot states:")
            # fmt: off
            print("{")
            print(f"timestamp: [{states.timestamp[0]}, {states.timestamp[1]}]")
            print(f"q: {['%.3f' % i for i in states.q]}")
            print(f"theta: {['%.3f' % i for i in states.theta]}")
            print(f"dq: {['%.3f' % i for i in states.dq]}")
            print(f"dtheta: {['%.3f' % i for i in states.dtheta]}")
            print(f"tau: {['%.3f' % i for i in states.tau]}")
            print(f"tau_des: {['%.3f' % i for i in states.tau_des]}")
            print(f"tau_dot: {['%.3f' % i for i in states.tau_dot]}")
            print(f"tau_ext: {['%.3f' % i for i in states.tau_ext]}")
            print(f"tau_interact: {['%.3f' % i for i in states.tau_interact]}")
            print(f"temperature: {['%.3f' % i for i in states.temperature]}")
            print(f"flange_pose: {['%.3f' % i for i in states.flange_pose]}")
            print(f"tcp_pose: {['%.3f' % i for i in states.tcp_pose]}")
            print(f"tcp_twist: {['%.3f' % i for i in states.tcp_twist]}")
            print(f"tcp_wrench: {['%.3f' % i for i in states.tcp_wrench]}")
            print(f"tcp_wrench_local: {['%.3f' % i for i in states.tcp_wrench_local]}")
            print(f"raw_tcp_wrench: {['%.3f' % i for i in states.raw_tcp_wrench]}")
            print(f"raw_tcp_wrench_local: {['%.3f' % i for i in states.raw_tcp_wrench_local]}")
            print(f"raw_ft_sensor: {['%.3f' % i for i in states.raw_ft_sensor]}")
            print("}", flush=True)
            # fmt: on

        # Print digital inputs
        logger.info("Digital inputs:")
        print(robot.digital_inputs())
        time.sleep(1)


def main():
    # Create an event to signal the thread to stop
    stop_event = threading.Event()

    # Program Setup
    # ==============================================================================================
    # Parse arguments
    argparser = argparse.ArgumentParser()
    argparser.add_argument(
        "robot_sn",
        help="Serial number of the robot to connect. Remove any space, e.g. Rizon4s-123456",
    )
    args = argparser.parse_args()

    # Define alias
    logger = spdlog.ConsoleLogger("Example")
    # Print description
    logger.info(
        ">>> Tutorial description <<<\nThis tutorial does the very first thing: check connection "
        "with the robot server and print received robot states.\n"
    )

    try:
        # RDK Initialization
        # ==========================================================================================
        # Instantiate robot interface
        robot = flexivrdk.Robot(args.robot_sn)

        # Clear fault on the connected robot if any
        if robot.fault():
            logger.warn("Fault occurred on the connected robot, trying to clear ...")
            # Try to clear the fault
            if not robot.ClearFault():
                logger.error("Fault cannot be cleared, exiting ...")
                return 1
            logger.info("Fault on the connected robot is cleared")

        # Enable the robot, make sure the E-stop is released before enabling
        logger.info("Enabling robot ...")
        robot.Enable()

        # Wait for the robot to become operational
        while not robot.operational():
            time.sleep(1)

        logger.info("Robot is now operational")

    except Exception as e:
        # Print exception error message
        logger.error(str(e))
        return 1

    # Print States
    # =============================================================================
    # Thread for printing robot states
    print_thread = threading.Thread(
        target=print_robot_states, args=[robot, logger, stop_event]
    )
    print_thread.start()

    # Use main thread to catch keyboard interrupt and exit thread
    try:
        while not stop_event.is_set():
            time.sleep(0.1)
    except KeyboardInterrupt:
        # Send signal to exit thread
        logger.info("Stopping print thread")
        stop_event.set()

    # Wait for thread to exit
    print_thread.join()
    logger.info("Print thread exited")


if __name__ == "__main__":
    main()
